chia/plotting/cache.py

```diff
@@ -0,0 +1,176 @@
+import logging
+import time
+import traceback
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import Dict, ItemsView, KeysView, List, Optional, Tuple, ValuesView
+
+from blspy import G1Element
+from chiapos import DiskProver
+
+from chia.plotting.util import parse_plot_info
+from chia.types.blockchain_format.proof_of_space import ProofOfSpace
+from chia.types.blockchain_format.sized_bytes import bytes32
+from chia.util.ints import uint16, uint64
+from chia.util.path import mkdir
+from chia.util.streamable import Streamable, streamable
+from chia.wallet.derive_keys import master_sk_to_local_sk
+
+log = logging.getLogger(__name__)
+
+CURRENT_VERSION: int = 1
+
+
+@streamable
+@dataclass(frozen=True)
+class DiskCacheEntry(Streamable):
+    prover_data: bytes
+    farmer_public_key: G1Element
+    pool_public_key: Optional[G1Element]
+    pool_contract_puzzle_hash: Optional[bytes32]
+    plot_public_key: G1Element
+    last_use: uint64
+
+
+@streamable
+@dataclass(frozen=True)
+class DiskCache(Streamable):
+    version: uint16
+    data: bytes
+
+
+@streamable
+@dataclass(frozen=True)
+class CacheDataV1(Streamable):
+    entries: List[Tuple[str, DiskCacheEntry]]
+
+
+@dataclass
+class CacheEntry:
+    prover: DiskProver
+    farmer_public_key: G1Element
+    pool_public_key: Optional[G1Element]
+    pool_contract_puzzle_hash: Optional[bytes32]
+    plot_public_key: G1Element
+    last_use: float
+
+    @classmethod
+    def from_disk_prover(cls, prover: DiskProver) -> "CacheEntry":
+        (
+            pool_public_key_or_puzzle_hash,
+            farmer_public_key,
+            local_master_sk,
+        ) = parse_plot_info(prover.get_memo())
+
+        pool_public_key: Optional[G1Element] = None
+        pool_contract_puzzle_hash: Optional[bytes32] = None
+        if isinstance(pool_public_key_or_puzzle_hash, G1Element):
+            pool_public_key = pool_public_key_or_puzzle_hash
+        else:
+            assert isinstance(pool_public_key_or_puzzle_hash, bytes32)
+            pool_contract_puzzle_hash = pool_public_key_or_puzzle_hash
+
+        local_sk = master_sk_to_local_sk(local_master_sk)
+
+        plot_public_key: G1Element = ProofOfSpace.generate_plot_public_key(
+            local_sk.get_g1(), farmer_public_key, pool_contract_puzzle_hash is not None
+        )
+
+        return cls(prover, farmer_public_key, pool_public_key, pool_contract_puzzle_hash, plot_public_key, time.time())
+
+    def bump_last_use(self) -> None:
+        self.last_use = time.time()
+
+    def expired(self, expiry_seconds: int) -> bool:
+        return time.time() - self.last_use > expiry_seconds
+
+
+@dataclass
+class Cache:
+    _path: Path
+    _changed: bool = False
+    _data: Dict[Path, CacheEntry] = field(default_factory=dict)
+    expiry_seconds: int = 7 * 24 * 60 * 60  # Keep the cache entries alive for 7 days after its last access
+
+    def __post_init__(self) -> None:
+        mkdir(self._path.parent)
```

Comment:
`.create()` method?

Reply:
Hm, `__post_init__` seems simpler and leaves us with the constructor only, not the constructor which then doesn't `mkdir` + `create`.

Reply:
And now any time i want to do _anything_ ever with the class I have to create a directory on the disk.  And then hopefully I'll clean it up after.  This ends up being a philosophical question where do-nothing constructors combine with consistency to just never bother thinking 'is it ok in this case?'

Reply:
> And now any time i want to do _anything_ ever with the class I have to create a directory on the disk.

What could that _anything_ be with a class which is called `Cache` with the purpose to save/load to/from a path (which should exist on disk if the class is used) and manage cached items? :) I mean, if you ever want to change the purpose of the class you need to adjust it no matter if you have a `create()` or not so you can also simply adjust the `__post_init__` in this case, no?